core/trino-main/src/test/java/io/trino/sql/query/TestCheckConstraint.java

```diff
@@ -0,0 +1,370 @@
+                    throw new UnsupportedOperationException();
+                })
+                .withCheckConstraints(schemaTableName -> {
+                    if (schemaTableName.equals(new SchemaTableName("tiny", "nation"))) {
+                        return ImmutableList.of("regionkey < 10");
+                    }
+                    if (schemaTableName.equals(new SchemaTableName("tiny", "nation_multiple_column_constraint"))) {
+                        return ImmutableList.of("nationkey > 100 AND regionkey > 50");
+                    }
+                    if (schemaTableName.equals(new SchemaTableName("tiny", "nation_invalid_function"))) {
+                        return ImmutableList.of("invalid_function(nationkey) > 100");
+                    }
+                    if (schemaTableName.equals(new SchemaTableName("tiny", "nation_not_boolean_expression"))) {
+                        return ImmutableList.of("1 + 1");
+                    }
+                    if (schemaTableName.equals(new SchemaTableName("tiny", "nation_subquery"))) {
+                        return ImmutableList.of("nationkey > (SELECT count(*) FROM nation)");
+                    }
+                    if (schemaTableName.equals(new SchemaTableName("tiny", "nation_not_deterministic"))) {
+                        return ImmutableList.of("nationkey > random()");
+                    }
+                    throw new UnsupportedOperationException();
+                })
+                .withData(schemaTableName -> {
+                    if (schemaTableName.equals(new SchemaTableName("tiny", "nation"))) {
+                        return TPCH_NATION_DATA;
+                    }
+                    if (schemaTableName.equals(new SchemaTableName("tiny", "nation_multiple_column_constraint"))) {
+                        return TPCH_NATION_DATA;
+                    }
+                    if (schemaTableName.equals(new SchemaTableName("tiny", "nation_invalid_function"))) {
+                        return TPCH_NATION_DATA;
+                    }
+                    if (schemaTableName.equals(new SchemaTableName("tiny", "nation_not_boolean_expression"))) {
+                        return TPCH_NATION_DATA;
+                    }
+                    if (schemaTableName.equals(new SchemaTableName("tiny", "nation_subquery"))) {
+                        return TPCH_NATION_DATA;
+                    }
+                    if (schemaTableName.equals(new SchemaTableName("tiny", "nation_not_deterministic"))) {
+                        return TPCH_NATION_DATA;
+                    }
+                    throw new UnsupportedOperationException();
+                })
+                .build();
+
+        runner.createCatalog(MOCK_CATALOG, mock, ImmutableMap.of());
+
+        assertions = new QueryAssertions(runner);
+    }
+
+    @AfterAll
+    public void teardown()
+    {
+        assertions.close();
+        assertions = null;
+    }
+
+    /**
+     * @see #testMergeInsert()
```

Comment:
While reviewing the PR I was thinking whether it is the responsibility of the engine to verify that the constraints may imply not being able to insert any value in the table.

- constraint 1 : `regionkey > 1`  
- constraint 2: `regionkey <  1`

Is this a user responsibility or an engine responsibility to check such things?

Reply:
I'd say user

Reply:
I can't find that SQL standard prohibits such constraints. So, I suppose it's users' responsibility.

Reply:
@findinpath this is not a solvable problem in general